jupyterlab/labapp.py

```diff
@@ -586,6 +588,23 @@ def initialize_templates(self):
 
 
     def initialize_settings(self):
+        def hook(model, path, **kwargs):
+            if "content" in model and model["content"] is not None:
+                # content sent through HTTP, it must not be an RTC session
+                if path in ROOMS:
+                    raise web.HTTPError(409, "Document content cannot be present both in RTC session and HTTP request")
+                # keep the content sent through HTTP
+            else:
+                # no content sent through HTTP, it must be an RTC session
+                if path in ROOMS:
+                    # we found the RTC session as expected
+                    # set the document content from y-py
+                    model["content"] = ROOMS[path].get_source()
+                else:
+                    # RTC session not available, shouldn't happen
+                    raise web.HTTPError(410, "Could not find an RTC session corresponding to this document")
+
+        self.serverapp.contents_manager.pre_save_hook = hook
```

Comment:
Will this not conflict with other users of this hook?

Reply:
I.e. maybe we should try to create a more flexible hook system now while this is still fresh, so that we don't have jupyter_hub, lab, and X other extensions all fighting for the `pre_save_hook slot`.

Reply:
Good point, maybe jupyter_server could implement a setter for the hooks, that would append them to a list of hooks under the hood? All hooks of the list would then be called. That has the advantage of keeping the same API.
I can open a PR there if that seems reasonable.